IGC/WrapperLLVM/include/llvmWrapper/IR/ConstantFolder.h

```diff
@@ -31,11 +31,85 @@ namespace IGCLLVM
+        }
+        inline llvm::Constant* CreateSelect(llvm::Constant* C, llvm::Constant* True,
+            llvm::Constant* False) const override {
+            return m_baseConstantFolder.CreateSelect(C, True, False);
+        }
+
+        inline llvm::Constant* CreateGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::ArrayRef<llvm::Constant*> IdxList) const override {
+            return m_baseConstantFolder.CreateGetElementPtr(Ty, C, IdxList);
+        }
+
+        inline llvm::Constant* CreateGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::Constant* Idx) const override {
+            return m_baseConstantFolder.CreateGetElementPtr(Ty, C, Idx);
+        }
+
+        inline llvm::Constant* CreateGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::ArrayRef<llvm::Value*> IdxList) const override {
+            return m_baseConstantFolder.CreateGetElementPtr(Ty, C, IdxList);
+        }
+
+        inline llvm::Constant* CreateInBoundsGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::ArrayRef<llvm::Constant*> IdxList) const override {
+            return m_baseConstantFolder.CreateInBoundsGetElementPtr(Ty, C, IdxList);
+        }
+
+        inline llvm::Constant* CreateInBoundsGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::Constant* Idx) const override {
+            return m_baseConstantFolder.CreateInBoundsGetElementPtr(Ty, C, Idx);
+        }
+
+        inline llvm::Constant* CreateInBoundsGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::ArrayRef<llvm::Value*> IdxList) const override {
+            return m_baseConstantFolder.CreateInBoundsGetElementPtr(Ty, C, IdxList);
+        }
+#else
+        inline llvm::Value* FoldAdd(llvm::Value* LHS, llvm::Value* RHS, bool HasNUW = false,
+            bool HasNSW = false) const override {
+            return m_baseConstantFolder.FoldAdd(LHS, RHS, HasNUW, HasNSW);
+        }
+
+        inline llvm::Value* FoldAnd(llvm::Value* LHS, llvm::Value* RHS) const override {
+            return m_baseConstantFolder.FoldAnd(LHS, RHS);
+        }
+
+        inline llvm::Value* FoldOr(llvm::Value* LHS, llvm::Value* RHS) const override {
+            return m_baseConstantFolder.FoldOr(LHS, RHS);
+        }
+
+        inline llvm::Value* FoldICmp(llvm::CmpInst::Predicate P, llvm::Value* LHS, llvm::Value* RHS) const override {
+            return m_baseConstantFolder.FoldICmp(P, LHS, RHS);
+        }
+
+        inline llvm::Value* FoldSelect(llvm::Value* C, llvm::Value* True, llvm::Value* False) const override {
+            return m_baseConstantFolder.FoldSelect(C, True, False);
+        }
+
+        inline llvm::Value* FoldGEP(llvm::Type* Ty, llvm::Value* Ptr, llvm::ArrayRef<llvm::Value*> IdxList,
+            bool IsInBounds = false) const override {
+            return m_baseConstantFolder.FoldGEP(Ty, Ptr, IdxList, IsInBounds);
+        }
+#endif
```

Comment:
Can you summarize what's going on here?

Reply:
> Can you summarize what's going on here?

In LLVM 14, they have rewritten the whole `llvm::IRBuilderFolder` class. So the previous overloading no longer works due to renamed member functions.

Reply:
The way the code is right now eg. `CreateOr()` wouldn't be available in LLVM >= 14 would it?
The wrapper idea is to have one symbol in core IGC code that proxies implementation depending on LLVM version. If `Create*` functions are for LLVM < 14, and `Fold*` for LLVM >= 14 that misses the point because you would still need `#if LLVM` guards in code depending on what LLVM you are linking against.

Does `CreateAdd()` work the same as `FoldAdd()`? Is so the wrapper may be written like this:

```cpp
T fold*(...)
{
#if LLVM 14
  ret llvm::fold*(...)
#else
  ret llvm::create*(...)
#endif
}
```

Reply:
Upon close inspection, isn't this class actually a wrapper itself to just proxy the function calls to the wrapped base class?

> The wrapper idea is to have one symbol in core IGC code that proxies implementation depending on LLVM version. If `Create*` functions are for LLVM < 14, and `Fold*` for LLVM >= 14 that misses the point because you would still need `#if LLVM` guards in code depending on what LLVM you are linking against.

It does make sense to do this because I think the purpose of this class is to inherit a base class to avoid the `final` modifier?

Reply:
Yes, that's correct. Since some recent LLVM version the class became `final` and forced us to create this folder "proxy".

I see in the LLVM differential revisions, such as here: https://reviews.llvm.org/D117431, or here: https://reviews.llvm.org/D116968 that there's not much difference between certain `Create` and `Fold` versions, aside from additional type check. Some cases tough, like this change for GEPs: https://reviews.llvm.org/D117039 is much more nuanced and may warrant a separate PR just to accomodate it.

There are good and worse ways of doing this. I understand this gets complicated but the folder wrapper should ideally migrate to `Fold`s and implement these such that on LLVM < 14 we get `Create`s' functionality. This will cause more refactoring since every previous `Create` user will have to be changed to `Fold`, however I'm not seeing a better way to do this. Keeping both `Create`s and `Fold`s does not seem sensible, since LLVM didn't either, instead of deprecating. Having `Create`s or `Fold`s depending on LLVM version breaks API compatibility within IGC.

Reply:
> There are good and worse ways of doing this. I understand this gets complicated but the folder wrapper should ideally migrate to `Fold`s and implement these such that on LLVM < 14 we get `Create`s' functionality. This will cause more refactoring since every previous `Create` user will have to be changed to `Fold`, however I'm not seeing a better way to do this. Keeping both `Create`s and `Fold`s does not seem sensible, since LLVM didn't either, instead of deprecating. Having `Create`s or `Fold`s depending on LLVM version breaks API compatibility within IGC.

It seems strange because it did not give me any compile errors when I changed the code like this. Probably there is currently (as of writing) nothing is using `#ifdef`ed APIs.

I am not sure what I should do here. Should I add non-`override` variants of the newer APIs to the older LLVM version of the wrapper? Since removing `override` functions will cause the class to become a virtual class.

Reply:
> nothing is using #ifdefed APIs

The IGC repository is also a part of our proprietary drivers, so we cannot assume this.

> Should I add non-override variants of the newer APIs to the older LLVM version of the wrapper?

Ideally yes, assumming corresponding `Fold` is a renamed `Create`. Then change users of these functions to use `Fold`. The goal is to have one symbol to use regardless of underlying LLVM version.
This should work as expected:
```cpp
T Fold() const
#if LLVM 14
override {
    return m_base.Fold*();
}
#else
{
    return m_base.Create*();
}
#endif
```
https://godbolt.org/z/hPT8GTf7d

Reply:
Done.